connect-location/src/main/java/com/ifttt/location/BackupGeofenceMonitor.java

```diff
@@ -0,0 +1,210 @@
+                                    MonitoredGeofence.GeofenceState.Init
+                                )
+                            );
+                        }
+                        if (existingMap.containsKey(exitKey)) {
+                            map.put(id, existingMap.get(exitKey));
+                        } else {
+                            map.put(getExitFenceKey(id),
+                                new MonitoredGeofence(LocationEventUploader.EventType.Exit,
+                                    MonitoredGeofence.GeofenceState.Init
+                                )
+                            );
+                        }
+                        break;
+                    default:
+                        // No-op for other location types.
+                }
+            }
+        }
+
+        sharedPreferences.edit().putString(PREF_KEY_MONITORED_GEOFENCES, jsonAdapter.toJson(map)).apply();
+    }
+
+    /**
+     * Refresh the {@link MonitoredGeofence.GeofenceState} of a single geofence in the cached
+     * monitored geofences.
+     *
+     * @param fenceKey Geofence identifier, must match one of the geofences registered internally in
+     * the SDK. This identifier can be found in {@link com.ifttt.location.ConnectLocation.LocationStatusCallback#onLocationStatusUpdated(List)}.
+     * @param state New {@link MonitoredGeofence.GeofenceState} for the fence key.
+     */
+    void setState(String fenceKey, MonitoredGeofence.GeofenceState state) {
+        Map<String, MonitoredGeofence> map = getMonitoredGeofences();
+        MonitoredGeofence currentGeofence = map.get(fenceKey);
+        if (currentGeofence == null) {
+            return;
+        }
+
+        map.put(fenceKey, new MonitoredGeofence(currentGeofence.type, state));
+        sharedPreferences.edit().putString(PREF_KEY_MONITORED_GEOFENCES, jsonAdapter.toJson(map)).apply();
+    }
+
+    /**
+     * Read the cached monitored geofence given the fence key.
+     *
+     * @param fenceKey Geofence identifier, must match one of the geofences registered internally in
+     * the SDK. This identifier can be found in {@link com.ifttt.location.ConnectLocation.LocationStatusCallback#onLocationStatusUpdated(List)}.
+     * @return cached {@link MonitoredGeofence.GeofenceState} for the fence key, or null if state
+     * cannot be found.
+     */
+    @Nullable
+    MonitoredGeofence.GeofenceState getState(String fenceKey) {
+        Map<String, MonitoredGeofence> map = getMonitoredGeofences();
+        MonitoredGeofence currentGeofence = map.get(fenceKey);
+        return currentGeofence != null ? currentGeofence.state : null;
+    }
+
+    /**
+     * @return all monitored geofence states.
+     */
+    Map<String, MonitoredGeofence> getMonitoredGeofences() {
+        String monitoredGeofencesString = sharedPreferences.getString(PREF_KEY_MONITORED_GEOFENCES, null);
+        if (monitoredGeofencesString == null) {
+            return Collections.emptyMap();
```

Comment:
Why can't we let the function just return null instead?

Reply:
We could, but the intention is that we don't have to deal with null if we can return at least empty maps. I don't know if we need to differentiate between empty map and null.